### ISSUE TITLE
Which is the best technique for hierarchy input sampling?

### ISSUE THREAD

**Issue by marcrocasalonso** [jonathf/chaospy#229]
Dear Jonathan,

Frist of all, thank you in advance for your time.
I am Marc Rocas a user from your Chaospy package and I am doing my Industrial PhD. I have a question about which could be the best method for this example:
### _**Example:**_
Imagine a QoI=Y=f(X1,X2,..,X6) that depends on this 6 random parameters X1,X2,X3,X4,X5,X6. In this case _f_ is a black-box software. These 6 uncertainty parameters has normal distribution. X1=N(0,1), X2=N(0,1) until X6=N(0,1).
I want to create a hierarchy methodology where in each level I increase the number samples.
**Step 1 (Level 1):**
I generate 10 samples of the inputs X1, X2,X3,X4,X5,X6 with _for example by MC random sampling and I evaluate the model _f_ to obtain the _QoI._ After that, I evaluate some statistics with this 10 responses like the PDF, the mean, the St. Desv.
**Step 2 (Level 2):**
Now I generate 10 samples more. Then, I have 20 samples. The  previous 10 samples and the new 10 samples of this level. With this 20 samples I evaluate statistics and I compare the results with the step 1.

**Step 3 (Level 3):**
...

**Unitl Step L (Level _L):**
...

**_My quetion:_**
Which is the best method in terms of efficiency to generate samples of the input space X1, X2, X3,X4,X5,X6? In this case, I have tested with Random sampling (Inneficient to sample the input distributions). Another options is Hammersley, but is not usefull because when I create in the level 2 the new 10 samples I can not use the previous 10 samples of the level 1 because is not resampling the input distribution of the previous level.
My idea is using a sparse grid technique like Smolyak by _cp.generate_quadrature(Nnw,distribution,rule='G',sparse=True)_ because this method is considering the samples of each level when I increase the number of samples. But the problem here is that is inefficient when applied to rules that can not be nested.

**¿Which could be the best method for sampling the inputs if I want to use in each level the previous samples and the newest created?** The main idea is don't lose any sample because each evaluation(simulation) of the back-box model f needs 20 hours.

Thank you in advance for your time and your help.

Best regards, :)
Marc

**Comment by jonathf** [comment]
Assuming you know what level L roughly should be, you can easily pre-generate samples and use the first N samples to ensure nesting upwards using any approach. If you use anything except naive random, you don't even hae to store the samples between run, as they are all deterministic.

However, if you do not want to bind yourself to an upper L, I propose using `rule="halton"`, as it is assured to be nested backwards like you describe (I guess):
```python
>>> dist = cp.Iid(cp.Normal(), 2)

>>> dist.sample(3, rule="halton")
<<< 
array([[-1.15034938,  0.31863936, -0.31863936],
       [-0.1397103 ,  0.76470967, -0.76470967]])

>>> dist.sample(5, rule="halton")
<<< 
array([[-1.15034938,  0.31863936, -0.31863936,  1.15034938, -1.53412054],
       [-0.1397103 ,  0.76470967, -0.76470967,  0.1397103 ,  1.22064035]])
```

If you decide to use nested quadrature nodes (which is what Smolyak sparse grid depend on) with the Gaussian distribution, your options is a little bit limited. The schemes that are nested, usually works best for distributions close to Uniform. The only exception is perhaps `rule="leja"`, but that is slow and I am unsure how well it will scale into your 6-D problem.

TL;DR, For 20-hour black-box, I'd stick to Halton.

**Comment by marcrocasalonso** [comment]
Hi Jonathan,

Thank you for your response and opinion.
In which situation do you think it could be better to use nested quadrature nodes like smolyak with respecto to Halton or other reduction variance techniques?

Thank you for your quickly answer! I will make a test with Halton sequences
:)

**Comment by jonathf** [comment]
Personally, I usually recommend against using Smolyak sparsegrid when you don't have to. The method is more sensitive to the assumption that the model behaves as a polynomial. Small deviation, and you won't only get a wrong answer, you can easily get a very wrong answer.

But if you can assure that accuracy is high enough (from the problem description somehow) and you then want to maximize precision, then you use Smolyak.

**Comment by marcrocasalonso** [comment]
Thank you very much Jonathan for your opinion between Smolyak and Halton.
I am testing with Halton and I hope that it will work well.

Best regards,
Marc
